### PR TITLE
Feat/learner registration

### PR DESCRIPTION
**Pull request by evaj10**
## Change goal

- Add learner type for registration in order to distinguish basic and commercial learners
    - On learner registration, admin must chose one of the types (image 1)
    - On bulk learner registration, admin must chose one of the roles (image 2)
- Handle username uniqueness on registration
    - If response is 400, use MatSnackBar to show error message (image 3)
- Handle username uniqueness on bulk registration
    - Duplicate usernames are removed by default
    - Request is sent on 'Proveri' button to check if any of learners already exists
    - Given view when invalid data and existing learners are present (image 4)

## Resulting views
alt Register one form
alt Bulk register form
alt Error if username exists, status 400
alt View with invalid entries, existing and new learners

### CODE REVIEW

File: src/app/modules/management/model/bulk-accounts.model.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { CreateLearner } from "./create-learner.model";
+
+export interface BulkAccounts {
```
Comment (Luburic) [inline src/app/modules/management/model/bulk-accounts.model.ts:3]: maybe AccountRegistrationCollections ?

File: src/app/modules/management/stakeholders/learners/bulk-add/bulk-add.component.ts [hunk 2]
```diff
@@ -47,24 +61,46 @@ export class BulkAddComponent implements OnInit {
...
+      next: (data) => {
+        this.responseView = true;
+        this.responseErrorView = false;
+        const accounts = data as BulkAccounts;
+        if (accounts.newAccounts.length !== 0) this.newLearnersAdded = true;
+        this.existingLearners = new MatTableDataSource(accounts.existingAccounts);
+        this.newLearners = new MatTableDataSource(accounts.newAccounts);
+      },
+      error: (error) => {
+        if (error.error.detail.includes('Duplicate username')) this.responseErrorView = true;
```
Comment (Luburic) [inline src/app/modules/management/stakeholders/learners/bulk-add/bulk-add.component.ts:92]: The UI/UX for this is wonky. I would rework the general flow of bulk registration to the following:

1. Admin enters 1 to N accounts.
2. Admin clicks "Proveri".
3. The system displays three lists:
   - "Nalozi koji će biti kreirani" - this list should remove duplicates on the front-end. I would change the current responseErrorView to a snackbar and remove the HTML related to this in bulk-add.component.
   - "Unete stavke koje nemaju dobar format"
   - "Nalozi koji već postoje i neće biti ponovo kreirani"
4. The admin fixes issues or clicks "Save".

File: src/app/shared/generics/generic-form/generic-form.component.html [hunk 3]
```diff
@@ -32,6 +32,17 @@ <h3>{{label}}</h3>
       <mat-datepicker #picker></mat-datepicker>
       <mat-error *ngIf="formGroup.controls[field.code].invalid && formGroup.controls[field.code].touched">{{getErrorMessage(field.code)}}</mat-error>
     </mat-form-field>
+
+    <div *ngIf="field.type === 'radio'" class="margin-bottom">
```
Comment (Luburic) [inline src/app/shared/generics/generic-form/generic-form.component.html:36]: Nice.

**Review by Luburic — CHANGES_REQUESTED** [review]
Logic is fine, just need to tweak the UX if we agree on the changes.

One more important thing - we should disable the chatbot popup for commercialLearners.

**Comment by Luburic** [comment]
@evaj10 one more important thing - we should disable the chatbot popup for commercialLearners.

File: src/app/modules/layout/home/dashboard/course-card/course-card.component.html [hunk 4]
```diff
@@ -1,6 +1,6 @@
 <mat-card style="margin: 10px" fxLayout="column" fxFlex
     (click)="onCourseClick()"
-    [ngClass]="{'pointer': user.role === 'learner'}">
+    [ngClass]="{'pointer': user.role.includes('learner')}">
```
Comment (Luburic) [inline src/app/modules/layout/home/dashboard/course-card/course-card.component.html:3]: Are we sure that we covered all the locations where this type of logic exists?

File: src/app/modules/layout/home/dashboard/dashboard.component.html [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<h2 *ngIf="user?.role === 'learner' || user?.role === 'instructor'">Katalog kurseva</h2>
+<h2 *ngIf="user?.role.includes('learner') || user?.role === 'instructor'">Katalog kurseva</h2>
```
Comment (Luburic) [inline src/app/modules/layout/home/dashboard/dashboard.component.html:1]: Do we need to use "includes" for the instructor here as well?

File: src/app/modules/management/courses/courses.service.ts [hunk 6]
```diff
@@ -24,10 +24,4 @@ export class CoursesService {
   removeOwner(courseId: number, instructorId: number) {
     return this.http.delete(this.baseUrl + courseId + "/owners/" + instructorId);
   }
-
-  getLearners(indexes: string[]): Observable<PagedResults<StakeholderAccount>> {
-    // Should be moved/merged to LearnersService?
-    // Post because of potential URL length limit violation with query params
-    return this.http.post<PagedResults<StakeholderAccount>>(environment.apiHost + 'management/learners/selected', indexes);
```
Comment (Luburic) [inline src/app/modules/management/courses/courses.service.ts:31]: Nice housekeeping :)

File: src/app/shared/generics/model/field.model.ts [hunk 7]
```diff
@@ -1,9 +1,11 @@
 import {Crud} from './crud.model';
+import { FieldOption } from './field-option';
 
 export interface Field {
   code: string;
   type: string;
   label: string;
   required?: boolean;
   crud?: Crud;
+  options?: FieldOption[];
```
Comment (Luburic) [inline src/app/shared/generics/model/field.model.ts:10]: Can you briefly explain the intent with this field? How is it meant to be used in general?

**Review by Luburic — CHANGES_REQUESTED** [review]
One or two minor things to address and a conflict to resolve and you can merge.

File: src/app/modules/layout/home/dashboard/dashboard.component.html (hunk 5 above)
Comment (evaj10) [reply src/app/modules/layout/home/dashboard/dashboard.component.html:1]: No, instructor has only one role.

File: src/app/shared/generics/model/field.model.ts (hunk 7 above)
Comment (evaj10) [reply src/app/shared/generics/model/field.model.ts:10]: When creating fields that have multiple options (radio/checkbox), values and labels for given options can be configured separately. This way, user sees a meaningful label (eg. FTN) and on submit value that server understands is sent (eg. Regular).

**Pull request merged**